### PR TITLE
feat!: add memory limiting

### PR DESCRIPTION
**Pull request by ValeryAntopol**
This PR adds the memory limiting feature to the Marine runtime, for now only when it runs using wasmtime backend. A user should set a limit in the service config, then the runtime will refuse to allocate more memory for modules in the service if the resulting total memory consumed exceeds the limit. 

The memory limiter counts memory consumed by WebAssembly Memory and WebAssembly Table objects and nothing else.

The new config structure looks like this:
```
modules_dir = "artifacts/"
total_memory_limit = "128 MiB" # field for memory limit
# total_memory_limit = "Infinity" # a way to allow infinite memory growth

[[module]]
    name = "greeting"
    # max_heap_size = "10 KiB" # removed field
    logger_enabled = false
```

### CODE REVIEW

File: core/src/memory_statistic.rs [hunk 1]
```diff
@@ -18,7 +18,8 @@ use serde::Serialize;
 use serde::Deserialize;
 
 use std::fmt;
-use std::ops::Deref;
+//use std::ops::Deref;
```
Comment (monoid) [inline core/src/memory_statistic.rs:21]: To be removed.

File: crates/wasmtime-backend/src/store.rs [hunk 2]
```diff
@@ -43,12 +44,77 @@ pub struct WasmtimeContextMut<'s> {
...
+impl ResourceLimiter for MemoryLimiter {
+    fn memory_growing(
+        &mut self,
+        current: usize,
+        desired: usize,
+        _maximum: Option<usize>,
+    ) -> wasmtime::Result<bool> {
+        let grow_size = (desired - current) as u64;
+        let new_total_memory = self
+            .current_total_memory
```
Comment (monoid) [inline crates/wasmtime-backend/src/store.rs:91]: Is `new_total_memory` really different from `desired`?

File: marine/tests/memory_limiting.rs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+mod utils;
+
+use marine::{CallParameters, IValue, Marine, MarineError};
```
Comment (mikevoronov) [inline marine/tests/memory_limiting.rs:19]: let's follow our convention

File: core/src/memory_statistic.rs [hunk 4]
```diff
@@ -28,53 +28,39 @@ use std::ops::Deref;
...
-            match record.max_memory_size {
-                Some(max_memory_size) => {
-                    let max_memory_size = bytesize::ByteSize::b(max_memory_size as u64);
-                    writeln!(f, "{} - {}/{}", record.name, memory_size, max_memory_size)?;
-                }
-                None => {
-                    writeln!(f, "{} - {}", record.name, memory_size)?;
-                }
-            }
+            writeln!(f, "{} - {}", record.name, memory_size)?;
```
Comment (mikevoronov) [inline core/src/memory_statistic.rs:63]: let's print `allocation_rejects` also

File: crates/wasm-backend-traits/Cargo.toml [hunk 5]
```diff
@@ -10,6 +10,7 @@ edition = "2021"
 wasmer-it = { package = "wasmer-interface-types-fl", version = "0.27.0" }
 it-memory-traits = "0.4.0"
 
+indexmap = "2.0.2"
```
Comment (mikevoronov) [inline crates/wasm-backend-traits/Cargo.toml:13]: looks like stale code

File: crates/wasmtime-backend/Cargo.toml [hunk 6]
```diff
@@ -18,3 +18,4 @@ multimap = "0.8.3"
 paste = "1.0.14"
 anyhow = "1.0.75"
 log = "0.4.20"
+indexmap = "2.0.2"
```
Comment (mikevoronov) [inline crates/wasmtime-backend/Cargo.toml:21]: ditto

File: crates/wasmtime-backend/src/store.rs [hunk 7]
```diff
@@ -43,12 +44,75 @@ pub struct WasmtimeContextMut<'s> {
...
+            .checked_add(grow_size)
+            .expect("Total memory can never reach 2^64");
+
+        if new_total_memory > self.max_total_memory {
+            self.allocation_stats.allocation_rejects += 1;
```
Comment (mikevoronov) [inline crates/wasmtime-backend/src/store.rs:100]: I'd better introduce a new function with a name like `count_allocation_reject`

File: crates/wasmtime-backend/src/store.rs [hunk 8]
```diff
@@ -43,12 +44,75 @@ pub struct WasmtimeContextMut<'s> {
...
+            return Ok(false);
+        }
+
+        self.current_total_memory = new_total_memory;
+        Ok(true)
+    }
+
+    fn table_growing(
```
Comment (mikevoronov) [inline crates/wasmtime-backend/src/store.rs:108]: Let's count it also

File: marine/src/config/marine_config.rs [hunk 9]
```diff
@@ -78,6 +78,9 @@ pub struct MarineConfig<WB: WasmBackend> {
     /// Path to a dir where compiled Wasm modules are located.
     pub modules_dir: Option<PathBuf>,
 
+    /// Total memory available for the service (in bytes)
+    pub memory_limit: Option<u64>,
```
Comment (mikevoronov) [inline marine/src/config/marine_config.rs:82]: Let's call it also `total_memory_limit` to follow the config name

File: marine/tests/memory_limiting.rs [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+});
+
+static LIMIT_64_MIB: Lazy<marine::TomlMarineConfig> = Lazy::new(|| {
+    marine::TomlMarineConfig::load("./tests/wasm_tests/memory_limiting/64MiB_limit.toml")
+        .expect("toml faas config should be created")
+});
+const FACADE_MODULE: &str = "memory_limiting_pure";
+const KB: usize = 1024;
+const MB: usize = 1024 * KB;
+const WASM_PAGE_SIZE: usize = 64 * KB;
```
Comment (mikevoronov) [inline marine/tests/memory_limiting.rs:40]: let's use the `bytesize` crate which is used in Marine

**Review by mikevoronov — APPROVED** [review]
utACK

**Pull request merged**